javalib/src/main/scala/java/lang/Math.scala

```diff
@@ -1,324 +1,347 @@
-    else
-      (exp(a) + exp(-a)) / 2.0
+  @inline def getExponent(a: scala.Float): scala.Int =
+    cmath.ilogbf(a)
+
+  @inline def getExponent(a: scala.Double): scala.Long =
+    cmath.ilogb(a)
+
+  @inline def hypot(a: scala.Double, b: scala.Double): scala.Double =
+    cmath.hypot(a, b)
+
+  @inline def IEEEremainder(f1: scala.Double, f2: scala.Double): Double =
+    cmath.remainder(f1, f2)
+
+  @inline def incrementExact(a: scala.Int): scala.Int =
+    addExact(a, 1)
+
+  @inline def incrementExact(a: scala.Long): scala.Long =
+    addExact(a, 1L)
+
+  @inline def log(a: scala.Double): scala.Double =
+    `llvm.log.f64`(a)
+
+  @inline def log10(a: scala.Double): scala.Double =
+    `llvm.log10.f64`(a)
+
+  @inline def log1p(a: scala.Double): scala.Double =
+    log(a + 1)
+
+  @inline def max(a: scala.Double, b: scala.Double): scala.Double =
+    `llvm.maxnum.f64`(a, b)
+
+  @inline def max(a: scala.Float, b: scala.Float): scala.Float =
+    `llvm.maxnum.f32`(a, b)
+
+  @inline def max(a: scala.Int, b: scala.Int): scala.Int =
+    if (a > b) a else b
+
+  @inline def max(a: scala.Long, b: scala.Long): scala.Long =
+    if (a > b) a else b
+
+  @inline def min(a: scala.Double, b: scala.Double): scala.Double =
+    `llvm.maxnum.f64`(a, b)
+
+  @inline def min(a: scala.Float, b: scala.Float): scala.Float =
+    `llvm.minnum.f32`(a, b)
+
+  @inline def min(a: scala.Int, b: scala.Int): scala.Int =
+    if (a < b) a else b
+
+  @inline def min(a: scala.Long, b: scala.Long): scala.Long =
+    if (a < b) a else b
+
+  @inline def multiplyExact(a: scala.Int, b: scala.Int): scala.Int = {
+    val overflow = `llvm.smul.with.overflow.i32`(a, b)
+    if (overflow.flag) throw new ArithmeticException("Integer overflow")
+    else overflow.value
   }
 
-  def tanh(a: scala.Double): scala.Double = {
-    if (a.isNaN || a == 0.0)
-      a
-    else if (abs(a) == scala.Double.PositiveInfinity)
-      signum(a)
-    else {
-      // sinh(a) / cosh(a) =
-      // 1 - 2 * (exp(-a)/ (exp(-a) + exp (a)))
-      val expma = exp(-a)
-      if (expma == scala.Double.PositiveInfinity) //Infinity / Infinity
-        -1.0
-      else {
-        val expa = exp(a)
-        val ret  = expma / (expa + expma)
-        1.0 - (2.0 * ret)
-      }
-    }
+  @inline def multiplyExact(a: scala.Long, b: scala.Long): scala.Long = {
+    val overflow = `llvm.smul.with.overflow.i64`(a, b)
+    if (overflow.flag) throw new ArithmeticException("Long overflow")
+    else overflow.value
   }
 
-  def addExact(a: scala.Int, b: scala.Int): scala.Int = {
-    val res       = a + b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b < 0)) res
-    else throw new ArithmeticException("Integer overflow")
+  @inline def negateExact(a: scala.Int): scala.Int =
+    subtractExact(0, a)
+
+  @inline def negateExact(a: scala.Long): scala.Long =
+    subtractExact(0, a)
+
+  def nextAfter(a: scala.Float, b: scala.Double): scala.Float = {
```

Comment:
You should be able to use the same implementation as Scala.js: https://github.com/scala-js/scala-js/blob/v0.6.10/javalanglib/src/main/scala/java/lang/Math.scala#L134 although it is a bit [under-tested](https://github.com/scala-js/scala-js/blob/v0.6.10/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/MathTest.scala#L130).